cmd/hiffy/src/lib.rs

```diff
@@ -314,7 +466,61 @@ fn hiffy(
         };
 
         let op = idol::IdolOperation::new(hubris, func[0], func[1], task)?;
-        hiffy_call_print(hubris, core, &mut context, &op, &args)?;
+
+        // Very special-case handling: if someone didn't specify `--input`, but
+        // is piping data into the `humility` command, then we use `stdin` as
+        // the input source.
+        let input = if let Some(input) = subargs.input {
+            Some(std::fs::read(input)?)
+        } else if op.operation.leases.len() == 1
+            && op.operation.leases[0].read
+            && !op.operation.leases[0].write
+            && atty::isnt(atty::Stream::Stdin)
+        {
+            let mut v = vec![];
+            std::io::stdin().read_to_end(&mut v)?;
+            Some(v)
+        } else {
+            None
+        };
+
+        if let Some(input) = input {
+            let r = hiffy_call_write(
+                hubris,
+                core,
+                &mut context,
+                &op,
+                &args,
+                &input,
+            )?;
```

Comment:
I understand wanting to put the precondition checks inside hiffy_call_write but the error messages that are returned assume you know the internal details that --input invokes a method called hiffy_call_write.